profiler/src/ProfilerEngine/Datadog.Profiler.Native/ExceptionsProvider.cpp

```diff
@@ -0,0 +1,254 @@
+#include "ExceptionsProvider.h"
+
+#include "FrameStore.h"
+#include "HResultConverter.h"
+#include "Log.h"
+#include "OsSpecificApi.h"
+#include "shared/src/native-src/com_ptr.h"
+#include "shared/src/native-src/string.h"
+
+#define INVOKE(x)                                                                                             \
+    {                                                                                                         \
+        HRESULT hr = x;                                                                                       \
+        if (FAILED(hr))                                                                                       \
+        {                                                                                                     \
+            Log::Warn("Profiler call failed with result ", HResultConverter::ToStringWithCode(hr), ": ", #x); \
+            return false;                                                                                     \
+        }                                                                                                     \
+    }
+
+ExceptionsProvider::ExceptionsProvider(
+    ICorProfilerInfo4* pCorProfilerInfo,
+    IManagedThreadList* pManagedThreadList,
+    IFrameStore* pFrameStore,
+    IConfiguration* pConfiguration,
+    IAppDomainStore* pAppDomainStore,
+    IRuntimeIdStore* pRuntimeIdStore) :
+    CollectorBase<RawExceptionSample>(pConfiguration, pFrameStore, pAppDomainStore, pRuntimeIdStore),
+    _pCorProfilerInfo(pCorProfilerInfo),
+    _pManagedThreadList(pManagedThreadList),
+    _pFrameStore(pFrameStore),
+    _messageFieldOffset(),
+    _stringLengthOffset(0),
+    _stringBufferOffset(0),
+    _mscorlibModuleId(0),
+    _exceptionClassId(0)
+{
+    _pStackFramesCollector = OsSpecificApi::CreateNewStackFramesCollectorInstance(_pCorProfilerInfo, _pManagedThreadList);
```

Comment:
Today the collector is shared amongst threads that throw exceptions. But the collector has states (ex: the buffer where frames are stored) and we could end up in a race.
Maybe having 1 collector per core would prevent that. (or add a lock around its usage)

Reply:
I made the collector thread-safe. Maybe it would suffice for now.